File: kernel/src/arch/aarch64/board/armv8_timer.c
#include "board/armv8_timer.h"
#include "pimento.h"
#include "scheduler.h"

static unsigned _cntfrq = 0;

/**
 * Listen for ticks and delegate to the scheduler.
 */
void armv8_timer_tick(void * data)
{
    (void) data;

    schedule();
}

/**
 * Initialize the ARMv8 timer.
 */
void armv8_timer_init(void)
{
    asm volatile("mrs %0, cntfrq_el0" : "=r" (_cntfrq));
}

/**
 * Set the timer.
 */
void armv8_timer_set(unsigned ms)
{
    // Set down-count timer.
    asm volatile("msr cntp_tval_el0, %0" :: "r" (_cntfrq / 1000 * ms));

    // Enable the timer.
    asm volatile("msr cntp_ctl_el0, %0" :: "r" (1));
}

File: kernel/src/device/driver/armv8_timer.rs
use core::arch::asm;
use core::time::Duration;

use crate::device::{Monotonic, Timer};
use crate::sync::Lock;

const NANOS: u64 = 1_000_000_000;

/// AArch64 timer.
pub struct ArmV8Timer {
    lock: Lock,
}

impl ArmV8Timer {
    /// Creates a timer.
    pub unsafe fn new() -> Self {
        Self { lock: Lock::new() }
    }

    fn frequency(&self) -> u64 {
        // SAFETY: Safe because this is a read operation.
        unsafe {
            let cntfrq: u64;
            asm!("mrs {}, cntfrq_el0", out(reg) cntfrq);
            cntfrq
        }
    }

    fn count(&self) -> u64 {
        // SAFETY: Safe because this is a read operation.
        unsafe {
            let cntpct: u64;
            asm!("mrs {}, cntpct_el0", out(reg) cntpct);
            cntpct
        }
    }
}

impl Timer for ArmV8Timer {
    fn set_duration(&self, duration: Duration) {
        // SAFETY: Safe because call is behind a lock.
        self.lock.call(|| unsafe {
            // Set the timer.
            let tval = (duration.as_nanos() * self.frequency() as u128 / NANOS as u128) as u64;
            asm!("msr cntp_tval_el0, {}", in(reg) tval);

            // Enable the timer.
            asm!("msr cntp_ctl_el0, {}", in(reg) 1u64);
        });
    }
}

impl Monotonic for ArmV8Timer {
    fn monotonic(&self) -> Duration {
        let nanos = (self.count() as u128 * NANOS as u128 / self.frequency() as u128) as u64;
        Duration::from_nanos(nanos)
    }
}
